Calypso: Add transferred license ownership purchases to `me/purchases` page.

gmjuhasz opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Supersedes #101449 
Fixes MYJP-108

## Proposed Changes

* Prepared the dataviews view on the purchases page to show transferred licenses
* Did the same for the OG non-dataviews view
* Hid the actions for licenses that were transferred, as those are managed by the owner
* Added an info popup for those, and a link to support for assistance

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* If 176310-ghe-Automattic/wpcom is not merged yet, you have to check it out and sandbox the API
* Follow the instructiions on 172707-ghe-Automattic/wpcom to add a license and transfer it to another user
* Check `/me/purchases` as the user that originally bought the license, but then it was transfered to another user
* You should see the license in the list, but the manage button should not be there. It should also have a popup with the info, and a link to contact support
* Check the same page for the user that activated the license. They should be able to manage the license normally
* Do the same steps with the non-dataview view. To force it, you can return false for `useDataViewPurchasesList` in `client/me/purchases/controller.jsx` on L40.

![CleanShot 2025-06-10 at 11 00 47 png](https://github.com/user-attachments/assets/250cdd30-92ed-449a-a3d5-96d589a58622)




## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

elliottprogrammer committed:
Add transferred license purchases to `me/purchases` page.

gmjuhasz committed:
Merge branch 'trunk' into add/purchases-transferred-license-subscriptions

gmjuhasz committed:
Merge trunk

gmjuhasz committed:
Prepare Owner-info for transferred ownership

gmjuhasz committed:
Prepare purchase list

gmjuhasz committed:
Add support for dataviews view

gmjuhasz committed:
Merge branch 'trunk' into add/purchases-transferred-license-subscriptions-v2

gmjuhasz committed:
Merge branch 'trunk' into add/purchases-transferred-license-subscriptions-v2

gmjuhasz committed:
Copy change

robertsreberski commented:
bugbot run

gmjuhasz committed:
Fix optional chaning

gmjuhasz committed:
Cast purchaser ID to number

robertsreberski commented on client/lib/purchases/index.ts:
```diff
@@ -83,7 +83,9 @@ const siteStatus: SiteStatus = {
 	hasExpired: 30,
 	hasExpiringSoon: 40,
 	hasPaymentMethodExpired: 50,
+	sitelessPlan: 55, // i.e. - Akismet plans
```
It's a good change, just out of curiosity I wonder, why is it part of this PR?

gmjuhasz replied:
Do be fair I'm not sure it was added by Brian before I took it over

robertsreberski commented on client/lib/purchases/index.ts:
```diff
@@ -119,13 +126,20 @@ function getSitePurchasesStatus( site: SiteWithPurchases ) {
+function getPurchaseStatus( purchase: PurchaseWithStatus, userId?: number ) {
 	const expiry = moment( purchase.expiryDate );
 
+	if ( userId && purchase.userId !== userId && purchase?.purchaserId === userId ) {
```
there is no need for optional chaining in purchaserId - if it's null, it's gonna crash with userId :D also, I see below that we treat `purchase` as defined here, and indeed existing Typescript typing suggests the same.

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -48,13 +52,14 @@ export interface PurchasesListProps {
 	isUserBlocked: boolean;
 	availableSessions: number[];
 	siteId: number | null;
+	userId: number | undefined;
```
Why undefined instead of null?

Also, if undefined, then we can write `userId?: number` I suppose

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -69,59 +74,86 @@ function MembershipSubscriptions( {
 
-	if ( purchases && purchases.length ) {
-		content = <PurchasesDataViews purchases={ purchases } sites={ sites } />;
+	if ( allPurchases && allPurchasesLoaded && allPurchases.length ) {
```
Not a big deal, more of a nitpick, but allPurchases seem to be always defined, and allPurchases.length > 0 is only when allPurchasesLoaded is true. so from all these checks, we could just leave allPurchases.length.

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -69,59 +74,86 @@ function MembershipSubscriptions( {
+
+	const allPurchases = useMemo( () => {
+		if ( allPurchasesLoaded ) {
+			return [ ...purchases, ...transferredOwnershipPurchases ];
```
Previous code assumed `purchases` might be null, so if the code crashes, it will be here, cause spread operator would run on undefined.

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -69,59 +74,86 @@ function MembershipSubscriptions( {
+		purchases &&
+		! purchases.length &&
+		! subscriptions.length &&
+		transferredOwnershipPurchases &&
+		! transferredOwnershipPurchases.length
```
`transferredOwnershipPurchases` will also never be null

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -189,18 +217,17 @@ function PurchasesListDataView(
 	hasLoadedUserPurchasesFromServer: hasLoadedUserPurchasesFromServer( state ),
 	isFetchingUserPurchases: isFetchingUserPurchases( state ),
-	purchases: getUserPurchases( state ),
+	purchases: getUserPurchases( state ) ?? [],
```
🙌

robertsreberski commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -189,18 +217,17 @@ function PurchasesListDataView(
-} ) )(
-	withStoredPaymentMethods( localize( PurchasesListDataView ), { type: 'card', expired: true } )
-);
+	userId: getCurrentUserId( state ) ?? undefined,
```
what value do we try to escape from here?

robertsreberski reviewed (commented):
Some comments, but it does work good!

gmjuhasz committed:
Update transferred purchases API endpoint to /me/purchases/transferred

gmjuhasz committed:
Remove unnecessary optional chaining from purchaserId in getPurchaseStatus

gmjuhasz committed:
Improve TypeScript types and code robustness in purchases list

- Change userId type from 'number | undefined' to optional 'number?'
- Add null safety to purchases spread operator
- Simplify condition checks for allPurchases
- Remove unnecessary null checks for transferredOwnershipPurchases
- Use consistent null handling for getCurrentUserId

gmjuhasz commented:
Addressed PR feedback:

- Fixed optional chaining issue in afb76a96630  
- Fixed TypeScript types and robustness issues in b88e14b22bd
- Updated API endpoint to /me/purchases/transferred in 9834184d54a

All review comments have been addressed. Thanks @robertsreberski for the thorough review!

gmjuhasz committed:
Merge branch 'trunk' into add/purchases-transferred-license-subscriptions-v2

gmjuhasz committed:
Fix tooltip position

gmjuhasz committed:
Fix TypeScript type issues in purchases list DataViews

- Update userId type to accept null values from Redux state
- Handle null userId when calling useGetJetpackTransferredLicensePurchases hook

gmjuhasz committed:
Merge branch 'trunk' into add/purchases-transferred-license-subscriptions-v2

robertsreberski commented on client/me/purchases/purchase-item/index.tsx:
```diff
@@ -742,8 +743,13 @@ class PurchaseItem extends Component<
+			transferredOwnershipPurchases = [],
 		} = this.props;
 
+		const isTransferredOwnership = transferredOwnershipPurchases.some(
```
What would be the usual size of the purchases array here? If it's all transferred ownership purchases, we should probably consider the solution of Set.

robertsreberski commented on client/me/purchases/purchase-item/index.tsx:
```diff
@@ -809,16 +815,27 @@ class PurchaseItem extends Component<
 		let href;
 
-		if ( getManagePurchaseUrlFor && slug ) {
+		if ( getManagePurchaseUrlFor && slug && ! isTransferredOwnership ) {
```
Copilot got a good spot, again 😅 it's especially important if the calculation might get operation-expensive (the purchases array will grow to thousands) with time

robertsreberski commented on packages/data-stores/src/purchases/lib/assembler.ts:
```diff
@@ -124,6 +124,10 @@ export function createPurchaseObject( purchase: RawPurchase | RawPurchaseCreditC
 		isAutoRenewEnabled: parseInt( purchase.auto_renew ?? '' ) === 1,
 	};
 
+	if ( purchase.purchaser_id ) {
```
That is desired, AFAIK

robertsreberski commented on packages/data-stores/src/purchases/queries/use-get-jetpack-transferred-license-purchases.tsx:
```diff
@@ -0,0 +1,47 @@
+export interface PurchasesIndex {
+	[ purchaseId: number ]: Purchase;
+}
+
```
Interesting 🤔

robertsreberski reviewed (approved):
The code looks a lot better to me, I see all my previous comments addressed!

It is good to go generally, but Copilot seems to make couple of good nitpicks. They don't seem time consuming so I would recommend just addressing them.

gmjuhasz committed:
Optimize transferred ownership checks with Set-based lookup

- Replace O(n) array.some() with O(1) Set.has() for performance
- Create reusable utility in hooks/use-is-transferred-ownership.ts
- Update all components to use centralized logic
- Remove duplicate implementations across purchase components

gmjuhasz committed:
Avoid duplicate evaluation of isTransferredOwnership

Store result in local variable to prevent calling the function twice
for the same item in the render method

robertsreberski reviewed (approved):
It does good look now, kudos!

sirbrillig commented on client/me/purchases/purchase-item/owner-info.tsx:
```diff
@@ -1,23 +1,48 @@
+				"This license was activated on {{strong}}%(domain)s{{/strong}} by another user. If you haven't given the license to them on purpose, {{link}}contact our support team{{/link}} for more assistance.",
+				{
+					args: {
+						domain: purchase.domain || purchase.siteName || 'a site',
```
It seems like `'a site'` should be translated in case it's used. 

Alternatively you could provide a whole different translation string if `purchase.domain` and `purchase.siteName` are falsy.

robertsreberski replied:
Good spot, thanks!

sirbrillig commented on client/me/purchases/purchases-list-in-dataviews/index.tsx:
```diff
@@ -74,65 +79,87 @@ function MembershipSubscriptions( {
+
+	const allPurchases = useMemo( () => {
+		if ( allPurchasesLoaded ) {
+			return [ ...( purchases || [] ), ...transferredOwnershipPurchases ];
```
This works but I'm curious: could we just alter the purchases endpoint to return these purchases as well (and perhaps flag them with a special property)? Maybe it's not as easy as I imagine but it seems like it would greatly simplify all this logic and if we ever need to know the user's purchases in a different place then we'd have it without having to jump through extra hoops. Is there a concern that those purchases need to only be accessed explicitly?

robertsreberski replied:
That's a good question. This project has plenty of backstory, and was created by ex-a11n. I don't see any cons in extending /purchases endpoint instead. 🤔 

Maybe it just would slow down the endpoint response as it requires additional database query (176310-ghe-Automattic/wpcom). The idea is that there are not many purchases like this, I believe.

I do see how it would simplify the logic, even updating SA so that HEs would have easier visualization on what happened with the subscription.

Since the PR is functionally working, and adding the transferred licenses to /me/purchases list, I will go ahead and merge it. However, I am creating additional Linear issues to that project to investigate merging the endpoints, and extending SA functionality.

gmjuhasz replied:
I agree that we could do a spike on this to see if that solution is possible. This approach was created by Brian as I took over the PR. I do think both have their downsides

sirbrillig reviewed (commented):
I haven't tested this but does the individual subscription page linked from each purchase work correctly with this added data? Do the "Change payment method" and "Cancel" etc. actions work? Or is there something in here that prevents linking to the subscription page?

sirbrillig commented:
Also my question above was not answered,

> does the individual subscription page linked from each purchase work correctly with this added data? Do the "Change payment method" and "Cancel" etc. actions work? Or is there something in here that prevents linking to the subscription page?

I'll try to follow the instructions to see if I can figure it out myself, but if it's broken we may need to do some quick patching to fix things or revert this change.

sirbrillig commented:
I decided to revert in https://github.com/Automattic/wp-calypso/pull/104471 as the simplest solution to the above questions, primarily the untranslated strings. This PR can easily be recreated while we wait for the translations.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17575399).
